File: fixtures/scale/kotlin/large/src/cryptobench/util/Hex.kt
package cryptobench.util

object Hex {
    private const val DIGITS = "0123456789abcdef"

    fun encode(data: ByteArray): String {
        val sb = StringBuilder(data.size * 2)
        for (b in data) {
            sb.append(DIGITS[(b.toInt() shr 4) and 0xF]).append(DIGITS[b.toInt() and 0xF])
        }
        return sb.toString()
    }

    fun decode(hex: String): ByteArray =
        ByteArray(hex.length / 2) { i ->
            hex.substring(i * 2, i * 2 + 2).toInt(16).toByte()
        }
}

File: fixtures/scale/dart/large/lib/util/hex.dart
import 'dart:typed_data';

const _digits = '0123456789abcdef';

String hexEncode(Uint8List data) {
  final sb = StringBuffer();
  for (final b in data) {
    sb.write(_digits[(b >> 4) & 0xF]);
    sb.write(_digits[b & 0xF]);
  }
  return sb.toString();
}

Uint8List hexDecode(String hex) {
  final out = Uint8List(hex.length ~/ 2);
  for (var i = 0; i < out.length; i++) {
    out[i] = int.parse(hex.substring(i * 2, i * 2 + 2), radix: 16);
  }
  return out;
}
